packages/core/src/components/popover/popover.tsx

```diff
@@ -669,6 +671,23 @@ export class Popover extends AbstractComponent<IPopoverProps, IPopoverState> {
         }
     }
 
+    private setOpenStateDirectly(isOpen: boolean | ((previousIsOpen: boolean) => boolean)) {
+        if (typeof isOpen === "function") {
+            this.setState(previousState => {
```

Comment:
Do we need this callback approach? We don't really use it elsewhere in the codebase these days (for better or for worse). More concisely:

```tsx
private setOpenStateDirectly(isOpen: boolean | ((previousIsOpen: boolean) => boolean)) {
    const nextIsOpen = Utils.isFunction(isOpen) ? isOpen(this.state.isOpen) : isOpen;
    if (nextIsOpen) {
        this.checkDOMForDarkTheme();
    }
    this.setState({ isOpen: nextIsOpen });
});
```

EDIT: I wrote this before my comment on L590 above. If we can in fact do away with the function param support, that'd be preferable.

Reply:
Presumable the reason the original code has the `this.setState(state => ...` format is to avoid reading from `this.state` which can be potentially stale if there are multiple `setStates` prior to a `willUpdate` being called. (https://stackoverflow.com/questions/42038590/when-to-use-react-setstate-callback)

Reply:
That post is arguing for a callback as the _second_ parameter to `setState`. Above, we're providing a callback as the first parameter. Since we're only referring to the old `isOpen` value in the following lines, it's not clear to me that we need a callback at all.

Reply:
Sorry, I was mostly referencing the opening point "Yes there is, since setState works in an asynchronous way. That means after calling setState the this.state variable is not immediately changed". If some code path has recently called setState and changed the value of isOpen, before calling into this function we could potentially miss the update by doing a stale read. Happy to change it if that seems off.

Reply:
My understanding is that `state` changes may be batched until the next render cycle, but they won't be applied piecemeal before then (that would seem to violate React's fundamental lifecycle guarantees). I don't see any `setState` calls interleaved here, so I think we're fine not to use the callback.

@giladgray what say you?

Reply:
@cmslewis you're correct that `state` changes may be batched, which is precisely why the callback is necessary: by the time your `setState` is executed, the actual `state` value may have changed. the callback syntax allows you to respond to the exact latest value, **which is actually critical in this component** to ensure `isOpen` consistency at all times.

i believe i used the callback intentionally here to handle `isOpen` flickering, like if you click the button a whole bunch of times before one transition finishes.

Reply:
Sweet. 👍